src/v/storage/segment_index.h

```diff
@@ -197,6 +197,24 @@ class segment_index {
           _state.broker_timestamp, _state.max_timestamp, _retention_timestamp);
     }
 
+    // Check if compacted timestamp has a value.
+    bool has_clean_compact_timestamp() const {
+        return _state.clean_compact_timestamp.has_value();
+    }
+
+    // Set the compacted timestamp, if it doesn't already have a value.
+    void maybe_set_clean_compact_timestamp(model::timestamp t) {
```

Comment:
do we expect this method to be called always with the same timestamp? or that timestamps are in some order? shall we add an (debug) assert for the invariant?

Reply:
I only ever expect to set an index's `_clean_compact_timestamp` once, upon being marked clean by sliding window compaction. Any further calls/timestamps to this function should be ignored.

We could expect that any future timestamps passed to this function will be in increasing order, but I'm not sure we need to encode that invariant here given the behavior. What do you think?